MRG+1: nonuniform image option in tfr plot

mmagnuski opened the pull request:
fixes #3847
This is not finished API-wise. The way it works currently is that there is an `imtype` arg (temporary name - any ideas?) that allows `'normal'` or `'nonuniform'` and thus changes the plotting artist to be NonUniformImage.
I make it dependent on a kwarg because:
* NonUniformImage is a little less flexible - does have only two interpolation options for example ('none' and 'bilinear')
* I did not check it but it may be slower
* having a kwarg opens the door to adding `'logy'` option (next PR maybe - if you think that would be useful, see the last image)

A quick example:
* I used following `freqs`:
  <img width="431" alt="freqs used" src="https://cloud.githubusercontent.com/assets/8452354/21945486/ca96d53a-d9db-11e6-9222-1b6930038ad3.PNG">
* The first image is with `imtype='normal'` and therefore has incorrect y axis tick labels (alpha desync is in 10 Hz), the second one is with `imtype='nonuniform'` and has correct y axis tick labels etc.
  <img width="268" alt="nonuniform image" src="https://cloud.githubusercontent.com/assets/8452354/21945527/077d64b4-d9dc-11e6-9d5e-b21ebbae89f5.PNG">

For comparison, this is with log y (not a grat example but log y is frequently used in time-freq literature):
```python
fig = tfr_avg.plot([chan_ind], baseline=(-1.5, -0.5), mode='mean', show=False)
fig.axes[0].set_yscale('log')
extent = fig.axes[0].images[0].get_extent()
fig.axes[0].images[0].set_extent(extent[:2] + tuple(np.log10(extent[2:])))
```
<img width="282" alt="logscale" src="https://cloud.githubusercontent.com/assets/8452354/21945613/7e55bd70-d9dc-11e6-95c1-cfe81eb1591f.PNG">

mmagnuski commented on mne/time_frequency/tfr.py:
```diff
@@ -887,7 +887,8 @@ def __init__(self, info, data, times, freqs, nave, comment=None,
              cmap='RdBu_r', dB=False, colorbar=True, show=True,
-             title=None, axes=None, layout=None, verbose=None):
+             title=None, axes=None, layout=None, verbose=None,
+             imtype='normal'):
```
I know that verbose should be last but `imtype` is temporary - you may for example want me to remove it :)

mmagnuski commented on mne/viz/topo.py:
```diff
@@ -249,16 +249,31 @@ def _check_vlim(vlim):
+        img = ax.imshow(tfr[ch_idx], extent=extent, aspect="auto",
+                        origin="lower", vmin=vmin, vmax=vmax,
+                        picker=picker, cmap=cmap)
+    elif imtype == 'nonuniform':
```
there should a test if imtype has allowable str option, I will check it if kwarg is the way to go

mmagnuski commented on mne/viz/tests/test_topo.py:
```diff
@@ -153,5 +153,7 @@ def test_plot_tfr_topo():
                   vmin=0., vmax=14., show=False)
     tfr.plot([4], baseline=(None, 0), mode='ratio', show=False, title='foo')
+    tfr.plot([4], baseline=(None, 0), mode='ratio', show=False, title='foo',
+             imtype='nonuniform')
```
very simple "test" - but I am open to suggestions how to make it actually test something more useful

Eric89GXL commented:
> I make it dependent on a kwarg because:

There is a lot of speculation here, -1 on this. I'd assume YAGNI (and KISS on API) until we hit the use case, and that's when we'd add the arg. Regarding performance, it's mpl plotting (mpl is not known for speed) and I doubt it's a huge difference, so I'd say always use `nonuniform` for now. And if you need more than `bilinear` interp, you should probably custom plot.

Eric89GXL commented:
> For comparison, this is with log y (not a grat example but log y is frequently used in time-freq literature):

What does `nonouniformimage` do if you set `ax.set(yscale='log')`?

mmagnuski commented:
Ok, I agree on the kwarg, I can remove it. But sill it might be useful to be able to specify log-y (although 3 lines is not that much... :)). And NonUniformImage doesn't go well with log-y (just checked) - the y tick labels change but the image stays the same.

mmagnuski commented:
(BWT - talking about speed - are there chances of vispy gaining dev tracktion again?)

agramfort commented:
I agree with Eric. Just make it work by default with the most natural use case and don't introduce a new param.

mmagnuski commented:
@agramfort Ok, I will. This will, however, make log y axis not possible (it would be incorrect with NonUniformImage).

agramfort commented:
if you want a logy in your plot you should have used a log scale of freqs
in the TFR

so to me YAGNI

mmagnuski commented:
@agramfort That's exactly the case I mention - when you use log freqs and don't want them to display on linear scale as nonuniformimage does.

mmagnuski commented:
BTW - log freqs (or more generally nonuniform freqs) was the reason for using NonUniformImage.

mmagnuski commented:
It seems that NonUniformImage messes up interactive cmap:
```
======================================================================
ERROR: Test TFR plotting.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/mne-tools/mne-python/mne/time_frequency/tests/test_tfr.py", line 403, in test_plot
    fig.canvas.key_press_event('up')
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/matplotlib/backend_bases.py", line 1731, in key_press_event
    self.callbacks.process(s, event)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/matplotlib/cbook.py", line 527, in process
    proxy(*args, **kwargs)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/matplotlib/cbook.py", line 405, in __call__
    return mtd(*args, **kwargs)
  File "/home/travis/build/mne-tools/mne-python/mne/viz/utils.py", line 1452, in key_press
    self.mappable.set_cmap(cmap)
  File "/home/travis/miniconda/envs/test/lib/python2.7/site-packages/matplotlib/image.py", line 799, in set_cmap
    raise RuntimeError('Cannot change colors after loading data')
RuntimeError: Cannot change colors after loading data
```
what do you think?

Eric89GXL commented:
Argh. Would pcolormesh work in all use cases? You'd need to compute edges
from the centers but otherwise should probably work right?

mmagnuski commented:
I'll try to check today (btw I've read that pcolormesh is noticeably slower).

mmagnuski commented:
Changed to pcolormesh, should be ok now. And I was wrong about the speed, I was not aware that there are pcolor, pcolormesh and pcolorfast - and pcolor is the slow one.

mmagnuski commented:
This is now green and ready, let me know what you think.

Eric89GXL commented on mne/viz/topo.py:
```diff
@@ -250,15 +250,28 @@ def _imshow_tfr(ax, ch_idx, tmin, tmax, vmin, vmax, onselect, ylim=None,
+    time_lims = np.concatenate([[times[0] - time_diff[0]], times[:-1] +
+                               time_diff, [times[-1] + time_diff[-1]]])
+    freq_lims = np.concatenate([[freq[0] - freq_diff[0]], freq[:-1] +
+                               freq_diff, [freq[-1] + freq_diff[-1]]])
```
Nice.

...but this will break if there is only one frequency. Either raise an error in that case, or (what I would do) just plot from `freq - 0.5, freq + 05`. I think you can get this by just doing a conditional `freq_diff = [0.5] if len(freq) < 2 else np.diff(freq) / 2.` Same comment about `time_diff`.

mmagnuski replied:
Ah, right, I didn't consider one-timepoint or one freq-bin case, I'll fix it the way you suggest.

mmagnuski commented:
Ok, I'll change example(s) and whats_new in fixes section. Thanks @Eric89GXL!

mmagnuski commented:
Ok, I have:
* rebased and updated whats_new
* changed only one tutorial ([here](http://mne-tools.github.io/dev/auto_tutorials/plot_artifacts_correction_ica.html)) - I didn't find any other tutorial/example using tfr.plot. I changed `freqs` there from linear to log-spaced. If it looks cryptic, I can revert, once you see that plotting looks good.
* I have also cleaned up `_imshow_tfr` - mostly removing unnecessary or unused args:
  * arguments not used in `_imshow_tfr`:
    * `vline`
    * `hline`
  * arguments not used in tfr.plot(), which was the only one using _imshow_tfr
    * `ylim` - was passed as None, so ylim was not used in _imshow_tfr
    * `picker` - was defaulting to True in _imshow_tfr but tfr.plot() was calling it with picker=False (the default for any Artist)
    * `title` - was not passed from tfr.plot() to _imshow_tfr - tfr.plot() was setting the title itself

I hope the tests come back green, local tests were good.

jaeilepp commented:
I have also cleaned up _imshow_tfr - mostly removing unnecessary or unused args:
       arguments not used in _imshow_tfr:
           vline
           hline
       arguments not used in tfr.plot(), which was the only one using _imshow_tfr
           ylim - was passed as None, so ylim was not used in _imshow_tfr
           picker - was defaulting to True in _imshow_tfr but tfr.plot() was calling it with picker=False (the default for any Artist)
           title - was not passed from tfr.plot() to _imshow_tfr - tfr.plot() was setting the title itself

I believe many of these arguments were there to make it compatible with the unified way the pick callbacks are handled. Now when I run the tutorial and click on one the plots on ``power.plot_topo()``, I get ``TypeError: _imshow_tfr() got an unexpected keyword argument 'ylim'``.

jaeilepp commented:
There probably should also be a test for the clicks (unless this causes unreasonable trouble on the backend we use for the tests).

mmagnuski commented:
Hm, that makes sense. Are all the kwargs I removed necessary for pick callback handling? I can revert this and add a test if needed (here or elsewhere, however you prefer).

mmagnuski commented:
Ok, it seems that only `ylim` is needed for the callback, I'll fix that.

mmagnuski commented:
All green 💚 
I'm also happy to report that logy works well with pcolormesh. 😃 
@jaeilepp - do you want me to add a test for plot_topo clicking here?

Eric89GXL commented:
> do you want me to add a test for plot_topo clicking here?

If it's a few lines and a few minutes yes, please

mmagnuski commented:
Definitely few lines, but I didn't use _fake_click before and didn't know the architecture of plot_topo. :) At the beginning I thought that each channel has a separate axis but it seems there is one big axis for all of them. But it didn't take long to create a passing test :)
So this should be ready to go 🚀, thanks for your help and reviews!

Eric89GXL commented on mne/viz/topo.py:
```diff
@@ -247,39 +247,43 @@ def _check_vlim(vlim):
         if interactive_cmap:
             ax.CB = DraggableColorbar(cbar, img)
-    if title:
-        plt.title(title)
```
What happened to `title`?

mmagnuski replied:
As I mentioned before - I've cleaned up _imshow_tfr a bit.
`title` was not used - `tfr.plot()` sets the title itself after using _imshow_tfr (more precisely - suptitle).

Eric89GXL commented on tutorials/plot_sensors_time_frequency.py:
```diff
@@ -87,7 +87,8 @@
 
-freqs = np.arange(6, 30, 3)  # define frequencies of interest
+# define frequencies of interest (log-spaced)
+freqs = np.logspace(*np.log10([6, 35]), num=8)
```
The y-axis limits don't look great to me here:

https://3531-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_tutorials/plot_sensors_time_frequency.html

It seems like rather than ending at the frequency limits, they should end at the `pcolormesh` boundaries beyond those. It makes it clearer it's an imshow-like plot.

Eric89GXL replied:
I also think we should probably set the `yticks` to the frequencies themselves

mmagnuski replied:
Right, I didn't notice that. I thought pcolormesh will handle it right, but I'll set the ylim to the correct values.
Regarding `yticks` - what would you suggest? Checking number/location of ticks and relocating them to closest frequency bins? I'm afraid not everything will work smoothly with many non-uniform freqs (the tutorial has only 8 freq bins).

Eric89GXL commented:
Most will have linear or log frequencies and display using appropriate
axes. So I'd do something like `yticks = freqs[np.unique(np.linspace(0,
len(freqs) - 1, 20).round().astype(int))]`

mmagnuski commented:
The y limits are now correct.
I have also tested about 5 different tick placement methods and although I have some favourites I don't think all will agree (so probably better to consider tick placement in another PR).

Just to sketch:
your suggestion works ok for lin freqs (lower figure) but doesn't do well for log (upper figure; ignore the images - they're wrong so they don't match):
<img width="278" alt="yticks_01_25_freqs" src="https://cloud.githubusercontent.com/assets/8452354/22048228/d8f8054c-dd2b-11e6-8e7c-28d03fa91933.PNG">

Something that seems to work well for linear freqs is looking for whole numbers and subsampling if necessary (but it is horrible for log freqs even if looking for approximatelly whole):
<img width="471" alt="yticks_02_find_step_full" src="https://cloud.githubusercontent.com/assets/8452354/22048413/1e9a6fee-dd2d-11e6-8cf5-061f1d977de8.PNG">

so for log-freqs the best I found was just to take a freq step that divides the freq range into requested number of pieces and look for freq bins closes to `np.arange(fmin, fmax, step)`
<img width="445" alt="yticks_02_find_dist_logspace" src="https://cloud.githubusercontent.com/assets/8452354/22048515/d8024470-dd2d-11e6-9b9d-21168b7cab62.PNG">

I would not change the default matplotlib tick placement - at least not here (as it probably requires longer discussion, my choice would be for example to test for uniformity and go for whole numbers if freq values are linear and go for freq-step otherwise) - WDYT?

Eric89GXL commented:
The last thing you propose should be equivalent to what I did, except that you just chose 11 frequencies instead of 20:
```
yticks = freqs[np.unique(np.linspace(0, len(freqs) - 1, 11).round().astype(int))]
```
The `arange` with an extra "bump" will get you 11 total points. Is it not?

> your suggestion works ok for lin freqs (lower figure) but doesn't do well for log

Well it uses indices into data that are either linearly or log spaced, so they should work equally well as long as the corresponding plots are linear or log spaced, respectively. In other words, the top plots look bad because you are plotting log freqs on a linear scale (and 20 points was too many compared to your 110, so they get bunched up.

I'm surprised the one-liner doesn't work, but given that it's not actually that simple we can do this in a follow-up PR like you suggest.

Eric89GXL commented:
Correction, what you proposed last will be equivalent to mine for linear-spaced frequencies. For log spaced frequencies that are plotted properly (using log y axis), your formulation won't work well (it forces a linear spacing) but mine should.

jaeilepp commented on mne/viz/topo.py:
```diff
@@ -247,39 +247,47 @@ def _check_vlim(vlim):
 
-    extent = (tmin, tmax, freq[0], freq[-1])
     cmap, interactive_cmap = cmap
+    extent = (tmin, tmax, freq[0], freq[-1])
```
You're not using ``extent`` anywhere.

mmagnuski replied:
Yes, that's a leftover from nonuniformimage and imshow before, thanks for the catch.

mmagnuski commented:
Yes, you are right - it was pretty late in my part of earth and I might have been not thinking straight. :) But that last approach was specifically for log-freqs on linear y (default).
I thought you don't want logy - that's why I was trying out some other things. So, would you prefer automatic logy for non-uniform freqs?

Eric89GXL commented:
> So, would you prefer automatic logy for non-uniform freqs?

Only in the case it's clearly log. If you really want to get this right, I'd do `yscale = 'auto' (default) | 'linear' | 'log'`: `auto` mode would go to `log` spacing only if it's likely to be log:
```
ratio = freqs[1:] / freqs[:-1]
if len(freqs) > 0 and freqs[0] > 0 and np.allclose(ratio, ratio[0]):
    yscale = 'log'
else:
    yscale = 'linear'
ratio = ratio[0]
```
I'd also go with the method I proposed before for choosing the yticks, but that can be another PR.

It occurred to me that (in this PR) if `yscale == 'log'` you should use geometric mean instead of arithmetic mean for the `pcolormesh` bounds, and also set the "past-first-and-last frequency" limits based on an additional `ratio` multiplication as opposed to `step` addition.

mmagnuski commented:
> It occurred to me that (in this PR) if yscale == 'log' you should use geometric mean instead of arithmetic mean for the pcolormesh bounds, and also set the "past-first-and-last frequency" limits based on an additional ratio multiplication as opposed to step addition.
Are you sure? When I was setting the bounds and later setting y scale as log - everything looked ok. I thought it is adjusted automatically when changing from linear to log.

mmagnuski commented:
Oops

Eric89GXL commented:
Use a more drastic log spacing like freqs = [1, 10, 100] and log yscale and
you'll see the failing of arithmetic mean. The bounds will not land in the
visual midpoint, but geometric mean will

On Jan 18, 2017 08:07, "Mikolaj Magnuski" <notifications@github.com> wrote:

> Oops
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3905#issuecomment-273470540>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACQZXlWDWVA8qQauRbvP5qMNk3-rVNGHks5rTg5-gaJpZM4LjUCK>
> .
>

mmagnuski commented:
Ok, I'll fix that in the evening, thanks!

Eric89GXL commented on mne/viz/topo.py:
```diff
@@ -248,35 +248,65 @@ def _check_vlim(vlim):
-                               freq_diff, [freq[-1] + freq_diff[-1]]])
+
+    # the same for frequency - depending on whether yscale is log
+    if yscale is 'linear':
```
don't use `is` with strings, use `==`

Eric89GXL commented on mne/viz/topo.py:
```diff
@@ -248,35 +248,65 @@ def _check_vlim(vlim):
+        from scipy.stats import gmean
+        log_freqs = np.concatenate([[freq[0] / ratio[0]], freq,
+                                   [freq[-1] * ratio[0]]])
+        freq_lims = gmean(np.vstack([log_freqs[:-1], log_freqs[1:]]), axis=0)
```
you can save an import and just do `np.sqrt(log_freqs[:-1] * log_freqs[1:])`

also add a check that if `yscale == 'log'` that all frequencies are positive

Eric89GXL commented on mne/viz/topo.py:
```diff
@@ -248,35 +248,65 @@ def _check_vlim(vlim):
+    # test yscale
+    if len(freq) < 2:
+        yscale = 'linear'
+    elif yscale is not 'linear':
```
`!=`

Eric89GXL commented on mne/viz/topo.py:
```diff
@@ -248,35 +248,65 @@ def _check_vlim(vlim):
+        yscale = 'linear'
+    elif yscale is not 'linear':
+        ratio = freq[1:] / freq[:-1]
+    if yscale is 'auto':
```
`==`

mmagnuski commented:
(I've seen your comments, I'll adress then in a while :) )

I have added `yscale` kwarg, fixed bin boundaries for log scale and added tick setting.
A comparison of `freqs=[1, 10, 100]` with `yscale='log'` (upper) and `yscale='linear'` (lower):
<img width="306" alt="yticks_03" src="https://cloud.githubusercontent.com/assets/8452354/22083700/7e6c2502-ddcd-11e6-9603-2414abf14388.PNG">

Additionally, log-scale is turned off when only one freq is used (cause it doesn't make much sense/difference then).

Eric89GXL commented:
Those plots LGTM

mmagnuski commented:
I guess I should add info about `yscale` to whats_new...

mmagnuski commented on mne/viz/topo.py:
```diff
@@ -247,39 +247,76 @@ def _check_vlim(vlim):
+
+    # test yscale
+    if len(freq) < 2 or np.any(freq <= 0):
+        yscale = 'linear'
```
should I add a warning if log scale is requested but ignored due to the check above?

Eric89GXL replied:
For now I would give an error if log scale is selected and DC is there. We can relax to a warning later if we need to. If there is one freq it's fine

mmagnuski commented:
Ok, I have rebased, added error when freq[0] == 0 and a test for it as well as updated whats_new about yscale.
I was hitting a strange unicode encode error when testing - hope it passes here...

mmagnuski commented:
The error I get running `test_topo.py`:
```
Traceback (most recent call last):
  File "mne/viz/tests/test_topo.py", line 198, in <module>
    run_tests_if_main()
  File "c:\proj\src\mne-python\mne\utils.py", line 2216, in run_tests_if_main
    elapsed_name, peak_mem, peak_name))
  File "C:\Anaconda3\lib\encodings\cp852.py", line 19, in encode
    return codecs.charmap_encode(input,self.errors,encoding_map)[0]
UnicodeEncodeError: 'charmap' codec can't encode character '\u2022' in position 16: character maps to <undefined>
```

Eric89GXL commented:
That seems like an unrelated (windows-specific) bug that can be fixed in another PR. I think if you use `nosetests mne/tests/test_whatever.py` instead of running it it will work, i.e. it's a bug with `run_tests_if_main`

Eric89GXL commented:
Now [that's a nice lookin' plot](https://3563-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_tutorials/plot_sensors_time_frequency.html):

![plot](https://3563-1301584-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/_images/sphx_glr_plot_sensors_time_frequency_005.png)

LGTM, +1 for merge

mmagnuski commented:
Thanks @Eric89GXL! I'm also happy with the result. :)
@jaeilepp - you looked into this PR before, maybe you could review again? :)

jaeilepp commented on mne/time_frequency/tfr.py:
```diff
@@ -961,6 +961,10 @@ def plot(self, picks, baseline=None, mode='mean', tmin=None,
+        yscale : 'auto' (default) | 'linear' | 'log'
+            The scale of y (frequency) axis. 'linear' gives linear y axis,
+            'log' leads to log-spaced y axis and 'auto' detects if frequencies
+            are log-spaced and only then sets the y axis to 'log'.
```
``yscale`` should be exposed for ``plot_topo`` too.

mmagnuski commented:
@jaeilepp - yscale is now exposed at the level of plot_topo. The caveat is that plot_topo uses `_imshow_tfr_unified` to plot tfr images (using imshow) which I did not modify, so yscale is used only when user clicks on one of the images and _imshow_tfr is called. I won't have time to look into `_imshow_tfr_unified` (for example to merge it with `_imshow_tfr`) till thursday or friday. What do you say to merging this PR and I'll fix `plot_topo` -> `_plot_topo` -> `_imshow_tfr_unified` to use yscale in a next PR?

jaeilepp commented:
Yeah, I'm fine with that. +1 for MRG.

Eric89GXL commented:
Thanks @mmagnuski

mmagnuski commented:
Thanks @Eric89GXL, @jaeilepp! 🚀
